### PR TITLE
Add mTLS (client cert) registry authentication

### LINKED ISSUE #2129: Client TLS
Not a bug report. This is regarding closed ticket: https://github.com/GoogleContainerTools/kaniko/issues/1106/.

Looking at the related code: https://github.com/GoogleContainerTools/kaniko/blob/1395e4682f80bf4286513392220597dcca60ce1c/pkg/util/transport_util.go#L77

The `--registry-certificate` flag enables one to provide a root CA to validate the remote registry's server certificate.

In the case where the remote registry supports authorizing a client to push based on the client's certificate (e.g. via the certificate's distinguished name), this would require kaniko to support something like `--client-certificate-path=/path/to/my-npe-cert-and-key` (for example).

Has there been any consideration to implementing such a feature?

Thank you.

### PR DESCRIPTION
**Pull request by fernferret**
Fixes #2129 #1106 

**Description**

The docker documentation shows how to [enable mututal TLS (mTLS) authentication with client certificates](https://docs.docker.com/engine/security/certificates/). For registries that are protected with this, kaniko's executor and warmer now both support adding client certs via:

```
--registry-client-cert my.reg.com=/certs/cert.crt,/certs/key.key
```

The format is identical to the `--registry-certificate` flag, so client certificates can be specified for each registry. The format on the right side of the equal sign is always 2 comma split paths: `/path/to/cert.crt,/path/to/key.key` with the contents being PEM formatted certificate and key. These will be ingested directly into the [tls.LoadX509KeyPair](https://pkg.go.dev/crypto/tls#LoadX509KeyPair) function.

**Submitter Checklist**

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [unit tests](../DEVELOPMENT.md#creating-a-pr)
- [x] Adds integration tests if needed.

_See [the contribution guide](../CONTRIBUTING.md) for more details._

**Reviewer Notes**

- [ ] The code flow looks good. 
- [ ] Unit tests and or integration tests added.

**Release Notes**

- kaniko adds a new flag `--registry-client-cert` to enable mTLS authentication with client certs
  - the format is similar to `--registry-certificates`, allowing one client certificate/key pair per registry destination: `my.reg.com=/certs/cert.crt,/certs/key.key`
  - The client certificate must be `PEM` encoded

### CODE REVIEW

File: cmd/executor/cmd/root.go
Comment (glennakamura) [inline cmd/executor/cmd/root.go:218]: Seems like the closing quote and period is missing for consistency with `RegistriesCertificates`.
```suggestion
	RootCmd.PersistentFlags().VarP(&opts.RegistriesClientCertificates, "registry-client-cert", "", "Use the provided client certificate for mutual TLS (mTLS) communication with the given registry. Expected format is 'my.registry.url=/path/to/client/cert,/path/to/client/key'.")
```

File: cmd/warmer/cmd/root.go
Comment (glennakamura) [inline cmd/warmer/cmd/root.go:89]: Seems like the closing quote and period is missing for consistency with `RegistriesCertificates`.
```suggestion
	RootCmd.PersistentFlags().VarP(&opts.RegistriesClientCertificates, "registry-client-cert", "", "Use the provided client certificate for mutual TLS (mTLS) communication with the given registry. Expected format is 'my.registry.url=/path/to/client/cert,/path/to/client/key'.")
```

File: pkg/util/transport_util_test.go
Comment (glennakamura) [inline pkg/util/transport_util_test.go:114]: A valid `"cert,key"` pair should be provided to isolate the non-matching registry name.
```suggestion
			opts: config.RegistryOptions{RegistriesClientCertificates: map[string]string{fmt.Sprintf("other.%s", registryName): "/path/to/client/certificate.cert,/path/to/client/key.key"}},
```

**Review by glennakamura — CHANGES_REQUESTED** [review]
Thank you.  I suggested some minor changes but I am successfully using this code to authenticate to a repository that requires client certificate authentication.

File: cmd/executor/cmd/root.go
Comment (fernferret) [reply cmd/executor/cmd/root.go]: Good catch, fixed!

File: cmd/warmer/cmd/root.go
Comment (fernferret) [reply cmd/warmer/cmd/root.go]: Good catch, fixed!

File: pkg/util/transport_util_test.go
Comment (fernferret) [reply pkg/util/transport_util_test.go]: Thanks, another good catch. This didn't fail the tests because this was just testing that the certs get passed to the specified registry (and not all or an incorrect registry), so it didn't really matter what was in here.

However if someone was reading the code they'd have seen it incorrect and it would have certainly raised questions, so this is a very good catch, thanks for the review!

**Comment by fernferret** [comment]
Hey @glennakamura thanks for the review, I think I addressed all of your findings.

I also pulled in the main branch to make sure there are no budding merge conflicts.

File: pkg/util/transport_util.go
Comment (imjasonh) [inline pkg/util/transport_util.go:101]: I think this should be an error, rather than logging and otherwise ignoring the invalid cert/key pair.

**Review by imjasonh — APPROVED** [review]

File: pkg/util/transport_util.go
Comment (fernferret) [reply pkg/util/transport_util.go:101]: Hi @imjasonh, I was trying to match the style/behavior of the certificate authority flag (`--registry-certificate`) when I made this a warning.

After writing this and looking at it now, I agree with you that this should be a fatal error; someone said "please use these certs" and it didn't find them.

Just to confirm, you were requesting this to be a fatal error and exit with an error code if the certificate is not loaded?

Comment (imjasonh) [reply pkg/util/transport_util.go:101]: Yes that's correct. If the flag is malformed (e.g., `registry.com=/path/to/cert,/path/to/key,/some/other/path`) or if the correctly specified file(s) can't be loaded, these should be fatal errors.

Same with registry certs above, now that you mention it, for consistency.

It's likely (hopefully) that not being able to load the certs will cause executions to fail later anyway, but it'd be better to fail as soon as we tell something is wrong here, instead of letting the build proceed and fail later (again, hopefully) if the necessary certs aren't available.

Comment (fernferret) [reply pkg/util/transport_util.go]: Hi @imjasonh,

I think this looks a lot better now, since the error is stated immediately. I think it's a lot easier to identify potential issues.

Here is the before/after (with a simple typo of omitting the `m` in `pem`):

## Before

```
docker run ... kaniko-debug ... --registry-certificate test.example.com=/certs/ca.pe
WARN[0000] Failed to load certificate /certs/ca.pe for test.example.com
  error="open /certs/ca.pe: no such file or directory"
error checking push permissions -- make sure you entered the correct tag name,
  and that you are authenticated correctly, and try again:
  checking push permission for "test.example.com/example/example:latest":
  creating push check transport for test.example.com failed:
  Get "https://test.example.com/v2/": x509: certificate signed by unknown authority
```

## After this PR

```
docker run ... kaniko-debug ... --registry-certificate test.example.com=/certs/ca.pe
error checking push permissions -- make sure you entered the correct tag name,
  and that you are authenticated correctly, and try again:
  making transport for registry "test.example.com":
  failed to load certificate /certs/ca.pe for test.example.com:
  open /certs/ca.pe: no such file or directory
```

I did resolve this for both the verification (authority) and client certs.

WDYT?

**Comment by fernferret** [comment]
Hi all, are there any other changes requested to get this merged? I'll keep merging with main to keep things fresh but I don't believe I have any other action items yet!

**Review by imjasonh — APPROVED** [review]

**Comment by imjasonh** [comment]
> Hi all, are there any other changes requested to get this merged? I'll keep merging with main to keep things fresh but I don't believe I have any other action items yet!

Sorry for the delay. This looks good to me, I'll merge when it passes tests.

Thanks for adding this! 👍

**Comment by fernferret** [comment]
Hi @imjasonh apologies for the test failure.

I didn't realize the full test suite didn't work with go 1.18/1.19. I had run the unit tests (the go ones) myself but hadn't used `make test` (my bad). When I reverted to 1.17 I was able to run `make test` locally and get a full green run this time!

Do you mind pushing another test run through for me?

**Comment by martonivan** [comment]
Hi! I'm just curious when this pull request is expected to be merged. The tests are passed as far as I see, changes are approved and the feature is highly desired. :)

**Comment by djmcgreal-cc** [comment]
@imjasonh ?

**Comment by classyrancor** [comment]
@imjasonh @glennakamura Any change this could get merged in? In our environment we need mTLS support and this would be extremely useful.

**Comment by obeyler** [comment]
Hi, Do you plan to merge it soon ?

**Comment by joseph-arturia** [comment]
Hi @fernferret !
I'm using your tool in order to push images to a harbor registry. I mounted a volume with the needed certificates and key but i get the following error during the build-and-push step: 

error checking push permissions -- make sure you entered the correct tag name, and that you are authenticated correctly, and try again: checking push permission for "[my.harbor.registry.net/development/test](http://my.habror.registry.net/development/test)": POST https://my.harbor.registry.net/v2/development/test/blobs/uploads/: UNAUTHORIZED: unauthorized to access repository: development/test, action: push: unauthorized to access repository: development/test, action: push

Step failed

Does it mean that i'm not authenticated or is the problem related to my harbor config ?

By the way, do you know if this is gonna be merged? it would be very useful !

**Comment by imjasonh** [comment]
> Hi, Do you plan to merge it soon ?

I'm not actively contributing to Kaniko.

@chuangw6 might know if someone at Google is interested in reviewing and merging this PR.

**Comment by classyrancor** [comment]
> Hi @fernferret ! I'm using your tool in order to push images to a harbor registry. I mounted a volume with the needed certificates and key but i get the following error during the build-and-push step:
> 
> error checking push permissions -- make sure you entered the correct tag name, and that you are authenticated correctly, and try again: checking push permission for "[my.harbor.registry.net/development/test](http://my.habror.registry.net/development/test)": POST https://my.harbor.registry.net/v2/development/test/blobs/uploads/: UNAUTHORIZED: unauthorized to access repository: development/test, action: push: unauthorized to access repository: development/test, action: push
> 
> Step failed
> 
> Does it mean that i'm not authenticated or is the problem related to my harbor config ?
> 
> By the way, do you know if this is gonna be merged? it would be very useful !

This means kaniko failed to authenticate against your Harbor.

**Comment by obeyler** [comment]
> This means kaniko failed to authenticate against your Harbor.


This harbor repository is public (just protected by MTLS) we can push on it with docker Cli without been logged so on we still not understand why kaniko need to be authenticated

**Comment by fernferret** [comment]
> > This means kaniko failed to authenticate against your Harbor.
> 
> @classyrancor @fernferret This harbor repository is public (just protected by MTLS) we can push on it with docker Cli without been logged so on we still not understand why kaniko failed to authenticate as it shouldn't have to be authenticated

Hi @obeyler I'm happy to have a look, but we should probably move a question like this to the [kaniko slack channel](https://kubernetes.slack.com/messages/CQDCHGX7Y/) to keep this PR discussion focused on getting this PR merged.

**Comment by obeyler** [comment]
👍

**Comment by djmcgreal-cc** [comment]
@chuangw6 - any chance of taking a look? You're down as a pending reviewer.

**Comment by chuangw6** [comment]
> > Hi, Do you plan to merge it soon ?
> 
> I'm not actively contributing to Kaniko.
> 
> @chuangw6 might know if someone at Google is interested in reviewing and merging this PR.

Sorry for the delay. I used to manage the release of the project but did not contribute to the code base. @imjasonh Would you mind to take a look at this PR or suggest someone who is familiar with the changes? Thanks

**Comment by classyrancor** [comment]
@glennakamura Do you know anyone that could take a look at this PR?

**Comment by skype007** [comment]
@aaron-prindle @imjasonh

**Comment by aaron-prindle** [comment]
Thanks all for your patience here with this PR and sorry @skype007 for the delay in responding - recently added as kaniko project maintainer and am just going through backlogged PRs now.  In going through the PR and testing the flow here, everything LGTM.  Merging

**Comment by aaron-prindle** [comment]
@joseph-arturia the `gcr.io/kaniko-project/executor:latest` image won't be updated until the next kaniko release is cut (similar to last release `v1.9.2` here - https://github.com/GoogleContainerTools/kaniko/releases) .  I plan on cutting a kaniko release (`v1.10.0`) later this week so that will be when this PR is available at `kaniko-project/executor:latest`.  In the meantime you can use the CI/CD builds that are created at each PR merge - in this case `gcr.io/kaniko-project/executor:b4f3885a925d20a161232d6e92f6814f80a2d20d` to get the latest build at the moment of writing this

**Comment by aaron-prindle** [comment]
@joseph-arturia `v1.10.0` is released now, https://github.com/GoogleContainerTools/kaniko/releases/tag/v1.10.0
`gcr.io/kaniko-project/executor:latest` now has this change

**Pull request merged**
